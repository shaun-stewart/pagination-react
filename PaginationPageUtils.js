export const PREVIOUS_PAGE = "PREVIOUS";
export const NEXT_PAGE = "NEXT";
export const ELLIPSIS = "...";

/**
 * Helper method for creating a range of numbers
 * range(1, 5) => [1, 2, 3, 4, 5]
 */
export const range = (from, to, step = 1) => {
  let i = from;
  const currentRange = [];

  while (i <= to) {
    currentRange.push(i);
    i += step;
  }

  return currentRange;
};

/**
 * Page numbers include strings representing ellipsis, next page and previous page.
 * Let's say we have 10 pages and we set pageNeighbours to 2
 * Given that the current page is 6
 * The pagination control will look like the following:
 *
 *  < (1) ... {4 5} [6] {7 8} ... (10) >
 *
 * (x) => terminal pages: first and last page(always visible)
 * [x] => represents current page
 * {...x} => represents page neighbours
 * ... => ellipsis shown when additional page numbers are hidden
 */
export const fetchPageNumbers = (totalPages, pageNeighbours, currentPage) => {
  /**
   * totalNumbers: the total page numbers to show on the control (2*page neighbours + current page + 2 ellipsis
   * (if an ellipsis is unneeded, an additional number will be displayed in its place to maintain placing)
   * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
   */
  const totalNumbers = pageNeighbours * 2 + 3; // 3 => current page + 2 ellipsis
  const totalBlocks = totalNumbers + 2;

  if (totalPages > totalBlocks) {
    const leftBound = currentPage - pageNeighbours;
    const rightBound = currentPage + pageNeighbours;
    const beforeLastPage = totalPages - 1;

    const startPage = leftBound > 2 ? leftBound : 2;
    const endPage = rightBound < beforeLastPage ? rightBound : beforeLastPage;

    let pages = range(startPage, endPage);

    const pagesCount = pages.length;
    const singleSpillOffset = totalNumbers - pagesCount - 1;

    /**
     * leftSpill: has hidden pages to the left
     * rightSpill: has hidden pages to the right
     */
    const leftSpill = startPage > 2;
    const rightSpill = endPage < beforeLastPage;

    if (leftSpill && !rightSpill) {
      const extraPages = range(startPage - singleSpillOffset, startPage - 1);
      pages = [PREVIOUS_PAGE, 1, ELLIPSIS, ...extraPages, ...pages, totalPages];
    } else if (!leftSpill && rightSpill) {
      const extraPages = range(endPage + 1, endPage + singleSpillOffset);
      pages = [1, ...pages, ...extraPages, ELLIPSIS, totalPages, NEXT_PAGE];
    } else if (leftSpill && rightSpill) {
      pages = [
        PREVIOUS_PAGE,
        1,
        ELLIPSIS,
        ...pages,
        ELLIPSIS,
        totalPages,
        NEXT_PAGE
      ];
    }

    return pages;
  }

  return range(1, totalPages);
};
