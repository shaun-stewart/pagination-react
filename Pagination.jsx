import React, { Component } from "react";
import PropTypes from "prop-types";
import Icon from "../Icon";
import PaginationDots from "./PaginationDots";
import PaginationNumbers from "./PaginationNumbers";
import {fetchPageNumbers, range} from "./PaginationPageUtils";
import "../semantic-ui/definitions/pagination.less";

class Pagination extends Component {
  constructor(props) {
    super(props);
    const { totalRecords = null, perPageLimit, pageNeighbours, renderDots} = props;

    this.perPageLimit = typeof perPageLimit === "number" ? perPageLimit : 10;
    this.totalRecords = typeof totalRecords === "number" ? totalRecords : 0;
    this.renderDots = renderDots;

    this.pageNeighbours =
      typeof pageNeighbours === "number"
        ? Math.max(0, Math.min(pageNeighbours, 2))
        : 0;

    this.totalPages = Math.ceil(this.totalRecords / this.perPageLimit);

    this.state = { currentPage: 1 };
  }

  componentDidMount() {
    this.gotoPage(1);
  }

  gotoPage = page => {
    const { onPageChanged = f => f } = this.props;

    const currentPage = Math.max(0, Math.min(page, this.totalPages));

    const paginationData = {
      currentPage,
      totalPages: this.totalPages,
      perPageLimit: this.perPageLimit,
      totalRecords: this.totalRecords
    };

    this.setState({ currentPage }, () => onPageChanged(paginationData));
  };

  handleClick = (page, evt) => {
    evt.preventDefault();
    this.gotoPage(page);
  };

  handlePreviousPageClick = evt => {
    evt.preventDefault();
    const { currentPage } = this.state;

    this.gotoPage(currentPage - 1);
  };

  handleNextPageClick = evt => {
    evt.preventDefault();
    const { currentPage } = this.state;

    this.gotoPage(currentPage  + 1);
  };

  render() {
    if (!this.totalRecords) return null;

    if (this.totalPages === 1) return null;

    const { renderDots, totalPages, pageNeighbours, handleClick, handlePreviousPageClick, handleNextPageClick } = this;
    const { currentPage } = this.state;

    return( <div data-test="pagination"><ul className="pagination">
            {
                renderDots ? <PaginationDots pages={range(1, totalPages)}
                                             currentPage={currentPage}
                                             onClickHandler={handleClick}/>
                            :
                            <PaginationNumbers pages={fetchPageNumbers(totalPages, pageNeighbours, currentPage)}
                                               currentPage={currentPage}
                                               onClickHandler={handleClick}
                                               onPreviousPageClickHandler={handlePreviousPageClick}
                                               onNextPageClickHandler={handleNextPageClick}/>
            }
            </ul></div>
    );
  }
}

/*
totalRecords - indicates the total number of records to be paginated.

perPageLimit - indicates the number of records to be shown per page. If not specified, it defaults to 30 as defined in the constructor().

pageNeighbours - indicates the number of additional page numbers to show on each side of the current page.
The minimum value is 0 and the maximum value is 2.
If not specified, it defaults to 0 as defined in the constructor().

onPageChanged - is a function that will be called with data of the current pagination state only when the current page changes.

renderDots - whether navigation should be displayed as dots (as seen in swiping components) or standard numbered pages
 */
Pagination.propTypes = {
  totalRecords: PropTypes.number.isRequired,
  perPageLimit: PropTypes.number,
  pageNeighbours: PropTypes.number,
  onPageChanged: PropTypes.func.isRequired,
  renderDots: PropTypes.bool
};

Pagination.defaultProps = {
  perPageLimit: 10,
  pageNeighbours: 0,
  renderDots: false
};

export default Pagination;
