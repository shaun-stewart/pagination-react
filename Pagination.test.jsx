import React from "react";
import renderer from "react-test-renderer";
import {mount} from "enzyme";
import Pagination from "./Pagination";
import {fetchPageNumbers, range, ELLIPSIS, PREVIOUS_PAGE, NEXT_PAGE} from "./PaginationPageUtils";



describe("Pagination", () => {
    it("should render pagination numbers properly", () => {
        const pagination = renderer
            .create(<Pagination totalRecords={15} perPageLimit={5} pageNeighbours={1} onPageChanged={()=>{}}/>)
            .toJSON();
        expect(pagination).toMatchSnapshot();
    });

    it("should render pagination dots properly", () => {
        const pagination = renderer
            .create(<Pagination totalRecords={15} perPageLimit={5} pageNeighbours={1} onPageChanged={()=>{}} renderDots/>)
            .toJSON();
        expect(pagination).toMatchSnapshot();
    });

    it("should default first number as active", () =>{
        const pagination = mount(
            <Pagination totalRecords={15} perPageLimit={5} pageNeighbours={1} onPageChanged={()=>{}}/>
        );
        expect(pagination.find('.page-item').length).toEqual(3);
        expect(pagination.find('.page-item').at(0).hasClass('active')).toEqual(true);
        expect(pagination.find('.page-item').at(2).hasClass('active')).toEqual(false);
    });

    it("should set selected number as active", () =>{
        const pagination = mount(
            <Pagination totalRecords={15} perPageLimit={5} pageNeighbours={1} onPageChanged={()=>{}}/>
        );
        expect(pagination.find('.page-item').length).toEqual(3);
        expect(pagination.find('.page-item').at(0).hasClass('active')).toEqual(true);
        expect(pagination.find('.page-item').at(2).hasClass('active')).toEqual(false);
        pagination.find('.page-link').at(2).simulate('click');
        expect(pagination.find('.page-item').at(0).hasClass('active')).toEqual(false);
        expect(pagination.find('.page-item').at(2).hasClass('active')).toEqual(true);

    });
});


describe("range method in pagination page utils", () => {
    it("should return correct range", () => {
       const returnedRange = range(1,10);
       expect(returnedRange.length).toBe(10);
       expect(returnedRange[0]).toBe(1);
       expect(returnedRange[5]).toBe(6);
       expect(returnedRange[9]).toBe(10);
    });
});

describe("fetchPageNumbers method in pagination page utils", () => {
    it("should return a range of (1 to total pages) if total pages is less that total blocks", () => {

        /**
         * totalNumbers: the total page numbers to show on the control (2*page neighbours + current page + 2 ellipsis
         * (if an ellipsis is unneeded, an additional number will be displayed in its place to maintain placing)
         * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
         */

        const totalPages = 3;
        const pageNeighbours = 1;
        const currentPage = 1;
        const pageNumbers = fetchPageNumbers(totalPages, pageNeighbours, currentPage);

        expect(pageNumbers.length).toBe(3);
        expect(pageNumbers[0]).toBe(1);
        expect(pageNumbers[2]).toBe(3);
    });



    it("should return correct page numbers, including next page and an ellipsis before last number, if total pages is greater that total blocks" +
        "  and current page + page neighbours < total pages - 1", () => {

        /**
         * totalNumbers: the total page numbers to show on the control (2*page neighbours + current page + 2 ellipsis
         * (if an ellipsis is unneeded, an additional number will be displayed in its place to maintain placing)
         * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
         */

        const totalPages = 15;
        const pageNeighbours = 1;
        const currentPage = 1;
        const pageNumbers = fetchPageNumbers(totalPages, pageNeighbours, currentPage);

        const expectedArray = [1,2,3,4,5, ELLIPSIS, 15, NEXT_PAGE];
        expect(pageNumbers.length).toBe(8);
        expect(pageNumbers).toStrictEqual(expectedArray);

    });

    it("should return correct page numbers, including previous page and an ellipsis after first number, if total pages is greater that total blocks" +
        "  and current page - page neighbours > 2", () => {

        /**
         * totalNumbers: the total page numbers to show on the control (2*page neighbours + current page + 2 ellipsis
         * (if an ellipsis is unneeded, an additional number will be displayed in its place to maintain placing)
         * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
         */

        const totalPages = 15;
        const pageNeighbours = 1;
        const currentPage = 15;
        const pageNumbers = fetchPageNumbers(totalPages, pageNeighbours, currentPage);

        const expectedArray = [PREVIOUS_PAGE,1,ELLIPSIS,11,12,13,14,15];
        expect(pageNumbers.length).toBe(8);
        expect(pageNumbers).toStrictEqual(expectedArray);

    });

    it("should return correct page numbers, including next and previous page and two ellipsis, if total pages is greater that total blocks" +
        " and current page + page neighbours < total pages - 1" +
        " and current page - page neighbours > 2", () => {

        /**
         * totalNumbers: the total page numbers to show on the control (2*page neighbours + current page + 2 ellipsis
         * (if an ellipsis is unneeded, an additional number will be displayed in its place to maintain placing)
         * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
         */

        const totalPages = 15;
        const pageNeighbours = 1;
        const currentPage = 7;
        const pageNumbers = fetchPageNumbers(totalPages, pageNeighbours, currentPage);

        const expectedArray = [PREVIOUS_PAGE,1,ELLIPSIS, 6,7,8, ELLIPSIS, 15, NEXT_PAGE];
        expect(pageNumbers.length).toBe(9);
        expect(pageNumbers).toStrictEqual(expectedArray);

    });
});