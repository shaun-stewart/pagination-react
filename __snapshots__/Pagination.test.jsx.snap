// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render pagination dots properly 1`] = `
<div
  data-test="pagination"
>
  <ul
    className="pagination"
  >
    <div>
      <li
        className="page-item active"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="icon icon-swipe-circle-active"
            onClick={[Function]}
          >
            <test-file-stub />
          </i>
        </a>
      </li>
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="icon icon-swipe-circle-inactive"
            onClick={[Function]}
          >
            <test-file-stub />
          </i>
        </a>
      </li>
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="icon icon-swipe-circle-inactive"
            onClick={[Function]}
          >
            <test-file-stub />
          </i>
        </a>
      </li>
    </div>
  </ul>
</div>
`;

exports[`Pagination should render pagination numbers properly 1`] = `
<div
  data-test="pagination"
>
  <ul
    className="pagination"
  >
    <div>
      <li
        className="page-item active"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          2
        </a>
      </li>
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/#"
          onClick={[Function]}
        >
          3
        </a>
      </li>
    </div>
  </ul>
</div>
`;
